core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSigner.java

```diff
@@ -0,0 +1,127 @@
+import software.amazon.awssdk.http.auth.aws.signer.BaseAwsV4HttpSigner;
+import software.amazon.awssdk.http.auth.spi.AsyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.AsyncSignedRequest;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignRequest;
+import software.amazon.awssdk.http.auth.spi.SyncSignedRequest;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * An implementation of a {@link AwsV4HttpSigner} that uses properties to compose
+ * v4-signers in order to delegate signing of a request and payload (if applicable) accordingly.
+ */
+@SdkInternalApi
+public final class DefaultAwsV4HttpSigner implements AwsV4HttpSigner {
+
+    private static final Logger LOG = Logger.loggerFor(DefaultAwsV4HttpSigner.class);
+
+    /**
+     * Given a request with a set of properties, determine which signer to delegate to, and call it with the request.
+     */
+    private static AwsV4HttpSigner getDelegate(
+        SignRequest<?, ? extends AwsCredentialsIdentity> signRequest) {
+
+        // create the base signer
+        BaseAwsV4HttpSigner<?> v4Signer = BaseAwsV4HttpSigner.create();
+        return getDelegate(v4Signer, signRequest);
+    }
+
+    /**
+     * Given a request with a set of properties and a base signer, compose an implementation with the base
+     * signer based on properties, and delegate the request to the composed signer.
+     */
+    public static AwsV4HttpSigner getDelegate(
+        BaseAwsV4HttpSigner<?> v4Signer,
+        SignRequest<?, ? extends AwsCredentialsIdentity> signRequest) {
+
+        // get the properties to decide on
+        String authLocation = validatedProperty(signRequest, AUTH_LOCATION, "Header");
+        Duration expirationDuration = validatedProperty(signRequest, EXPIRATION_DURATION, null);
+        boolean isPayloadSigning = validatedProperty(signRequest, PAYLOAD_SIGNING, true);
+        boolean isEventStreaming = validatedProperty(signRequest, EVENT_STREAMING, false);
+        Boolean isChunkedEncoding = validatedProperty(signRequest, CHUNKED_ENCODING, null);
+
+        if (authLocation.equals("Header")) {
+            v4Signer = new AwsV4HeaderHttpSigner((BaseAwsV4HttpSigner<AwsV4Properties>) v4Signer);
+        } else if (authLocation.equals("QueryString")) {
+            v4Signer = new DefaultAwsV4QueryHttpSigner((BaseAwsV4HttpSigner<AwsV4Properties>) v4Signer);
+            if (expirationDuration != null) {
+                v4Signer = new AwsV4PresignedHttpSigner((BaseAwsV4HttpSigner<AwsV4Properties>) v4Signer);
+            }
+        } else {
+            throw new IllegalArgumentException("Unrecognized auth-location option: " + authLocation);
+        }
+
+        if (isEventStreaming) {
+            v4Signer = loadEventStreamSigner(v4Signer);
+        }
+
+        if (isChunkedEncoding != null) {
```

Comment:
Why does this only check for non-null? As opposed to `true` similar to isEventStreaming? If someone sets this to false, they still get the S3 signer? Does S3 have operations where it would use `false`?

Reply:
>If someone sets this to false, they still get the S3 signer?

Also, the interface says this defaults to false (which makes sense), so if someone explicitly sets it to false v/s null, the behavior shouldn't be different.

Reply:
Because the property is strictly used to determine if it should use the event-stream signer or not (unlike the s3 case). Looking at `EnableChunkedEncodingInterceptor`, it appears that only put-object and upload-part (potentially) use chunked-encoding.